File: sasmodels/models/magnetic_langevin_core_shell_3D.c
  static double clipp(double value, double low, double high) //from kernel_iq.c
{
  return (value < low ? low : (value > high ? high : value));
} 

static double langevin(
    double x) {
    // cotanh(x)-1\x

    if (x < 0.00001) {
        // avoid dividing by zero
        return 1.0/3.0*x;
    } else {
        return 1.0/tanh(x)-1/x;
    }
}

static double langevinoverx(
    double x) {
    // cotanh(x)-1\x

    if (x < 0.00001) {
        // avoid dividing by zero
        return 1.0/3.0;
    } else {
        return langevin(x)/x;
    }
}

//from coremultishell.c



static double
outer_radius(double radius, double fp_n, double thickness[])
{
  double r = radius;
  int n = (int)(fp_n+0.5);
  for (int i=0; i < n; i++) {
    r += thickness[i];
  }
  return r;
}

static double
form_volume(double radius, double fp_n, double thickness[])
{
  return M_4PI_3 * cube(outer_radius(radius, fp_n, thickness));
}


effective_radius(int mode, double core_radius, double fp_n, double thickness[])
{
  switch (mode) {
  default:
  case 1: // outer radius
    return outer_radius(core_radius, fp_n, thickness);
  case 2: // core radius
    return core_radius;
  }
}


static double fqnucsq(double q, double sld_core, double radius,
   double sld_solvent, double fp_n, double sld[], double thickness[])
{
      const int n = (int)(fp_n+0.5);
  double f, r, last_sld;
  r = radius;
  last_sld = sld_core;
  f = 0.;
  for (int i=0; i<n; i++) {
    f += M_4PI_3 * cube(r) * (sld[i] - last_sld) * sas_3j1x_x(q*r);
    last_sld = sld[i];
    r += thickness[i];
  }
  f += M_4PI_3 * cube(r) * (sld_solvent - last_sld) * sas_3j1x_x(q*r);
    return 1e-4 *f*f;
}

static double fqMtranssq(double q, double sld_core, double radius,
   double sld_solvent, double fp_n, double sld[], double thickness[],double eta_core,double eta_solvent,double eta[], double delta_solvent,double delta[])
{
      const int n = (int)(fp_n+0.5);
  double f,fsq_perp, r, last_sld;
  r = radius;
  
  last_sld = sld_core*sqrt(langevinoverx(eta_core));
  f = 0.;
  fsq_perp=0;
  for (int i=0; i<n; i++) {
    f += M_4PI_3 * cube(r) *(sld[i]*sqrt(langevinoverx(eta[i])*delta[i]) - last_sld) * sas_3j1x_x(q*r);//transversal magnetisation component coaligned with core transversal magnetisation
    fsq_perp += square(M_4PI_3 * cube(r) *sld[i]*sqrt(langevinoverx(eta[i])*(1-delta[i])) * sas_3j1x_x(q*r));//transversal magnetisation component perpendicular with core transversal magnetisation, adding incoherently (only of squared amplitude)
    last_sld = sld[i]*sqrt(langevinoverx(eta[i])*delta[i]);   
    r += thickness[i];
  }
  f += M_4PI_3 * cube(r) * (sld_solvent*sqrt(langevinoverx(eta_solvent)*delta_solvent) - last_sld) * sas_3j1x_x(q*r);//transversal magnetisation component coaligned with core magnetisation
  fsq_perp += square(M_4PI_3 * cube(r) * sld_solvent*sqrt(langevinoverx(eta_solvent)*(1-delta_solvent))  * sas_3j1x_x(q*r));
    return 1e-4 *(f*f +fsq_perp);
}

static double fqMzsq(double q, double sld_core, double radius,
   double sld_solvent, double fp_n, double sld[], double thickness[],double eta_core,double eta_solvent,double eta[])
{
      const int n = (int)(fp_n+0.5);
  double f, r, last_sld;
  r = radius;
  last_sld = sld_core*sqrt(1-2*langevinoverx(eta_solvent));// sqrt() needed for correct scale to intensity later
  f = 0.;
  for (int i=0; i<n; i++) {
    f += M_4PI_3 * cube(r) * (sld[i]*sqrt(1-2*langevinoverx(eta_solvent)) - last_sld) * sas_3j1x_x(q*r);
    last_sld = sld[i]*sqrt(1-2*langevinoverx(eta[i]));
    r += thickness[i];
  }
  f += M_4PI_3 * cube(r) * (sld_solvent*sqrt(1-2*langevinoverx(eta_solvent)) - last_sld) * sas_3j1x_x(q*r);
    return 1e-4 *f*f;
}

static double fqnuc(double q, double sld_core, double radius,
   double sld_solvent, double fp_n, double sld[], double thickness[])
{
      const int n = (int)(fp_n+0.5);
  double f, r, last_sld;
  r = radius;
  last_sld = sld_core;
  f = 0.;
  for (int i=0; i<n; i++) {
    f += M_4PI_3 * cube(r) * (sld[i] - last_sld) * sas_3j1x_x(q*r);
    last_sld = sld[i];
    r += thickness[i];
  }
  f += M_4PI_3 * cube(r) * (sld_solvent - last_sld) * sas_3j1x_x(q*r);
    return 1e-2 *f;
}

static double fqMz(double q, double sld_core, double radius,
   double sld_solvent, double fp_n, double sld[], double thickness[],double eta_core,double eta_solvent,double eta[])
{
      const int n = (int)(fp_n+0.5);
  double f, r, last_sld;
  r = radius;
  last_sld = sld_core*langevin(eta_core);
  f = 0.;
  for (int i=0; i<n; i++) {
    f += M_4PI_3 * cube(r) * (sld[i]*langevin(eta[i]) - last_sld) * sas_3j1x_x(q*r);
    last_sld = sld[i]*langevin(eta[i]);
    r += thickness[i];
  }
  f += M_4PI_3 * cube(r) * (sld_solvent*langevin(eta_solvent) - last_sld) * sas_3j1x_x(q*r);
    return 1e-2 *f;
}




//Magnetization component wrt to field are fixed and defined by Langevin function.
//M_trans=M_x=M_y=0
//M_trans_quad=langevinoverx(xi)
//M_z= langevin(xi)
//M_zquad=(1-2*langevinoverx(xi))
//z axis oriented always along field/polarisation axis.



  
//some basic vector algebra

void SET_VEC(double *vector, double v0, double v1, double v2) {
    vector[0] = v0;
    vector[1] = v1;
    vector[2] = v2;
}

void SCALE_VEC(double *vector, double a) {
    vector[0] = a*vector[0];
    vector[1] = a*vector[1];
    vector[2] = a*vector[2];
}

void ADD_VEC(double *result_vec, double *vec1, double *vec2) {
    result_vec[0] = vec1[0] + vec2[0];
    result_vec[1] = vec1[1] + vec2[1];
    result_vec[2] = vec1[2] + vec2[2];
}



//Rotation of Detector around the magnetic reference frame
//Magnetic coordinates and scattering coordinates coincides when z/field axis along the neutron beam.
//then azimuthal angle theta=0Deg, rotation beta=arbitrary


static double square_mag_scat(
double cos_theta, double sin_theta,
    double alpha, double beta) {
    double cos_alpha, sin_alpha;
    double cos_beta, sin_beta;

    SINCOS(alpha*M_PI/180, sin_alpha, cos_alpha);
    SINCOS(beta*M_PI/180, sin_beta, cos_beta);
    //field is defined along (0,0,1), orientation of detector
    //is precessing in a cone around B with an inclination of theta
    double scat_vector[3];
    double perp_scat_vector[3];

    SET_VEC(scat_vector, cos_alpha* cos_theta, 
cos_theta* sin_alpha*sin_beta + 
 cos_beta* sin_theta, -cos_beta* cos_theta* sin_alpha + 
 sin_beta* sin_theta);



    return square(scat_vector[2]);

}


//weight of the magnetisation vector for the scattering, done by evaluating the magnetic scattering vector (Halpern Johnson vector) for general orientation of q. Inserting result in the spin-resolved scattering cross section
//and collecting terms 

//Calculation of the 4 spin-resolved scattering cross sections for superparamagnetic spheres, field perpendicular to beam 



//spin-resolved (POLARIS) cross sections
//NSF++ (F_N-Mz*(1-rotated_scat_vector[3]^2))^2+ Mx^2*(1-rotated_scat_vector[3]^2)*rotated_scat_vector[3]^2 
static double Idd(double q, double sld_core,double magnetic_sld_core,double eta_core,double radius,
   double sld_solvent,double magnetic_sld_solvent,double eta_solvent,double delta_solvent,
   double fp_n, double sld[],double magnetic_sld[],double eta[], double delta[], double thickness[], double cos_theta, double sin_theta, double alpha, double beta) { 
        
return fqnucsq(q, sld_core, radius, sld_solvent, fp_n, sld, thickness)-fqnuc(q, sld_core, radius, sld_solvent, fp_n, sld, thickness)*fqMz(q, magnetic_sld_core, radius, magnetic_sld_solvent, fp_n, magnetic_sld, thickness,eta_core,eta_solvent,eta)*(1-square_mag_scat(cos_theta, sin_theta,alpha, beta))+fqMzsq(q, magnetic_sld_core, radius, magnetic_sld_solvent, fp_n, magnetic_sld, thickness,eta_core,eta_solvent,eta)*square(1-square_mag_scat(cos_theta, sin_theta,alpha, beta))+fqMtranssq(q, magnetic_sld_core, radius, magnetic_sld_solvent, fp_n, magnetic_sld, thickness,eta_core,eta_solvent,eta, delta_solvent,delta)*(1-square_mag_scat(cos_theta, sin_theta,alpha, beta))*square_mag_scat(cos_theta, sin_theta,alpha, beta);
	}   
	
//NSF-- (F_N+Mz*(1-rotated_scat_vector[3]^2))^2+ Mx^2*(1-rotated_scat_vector[3]^2)*rotated_scat_vector[3]^2 
static double Iuu(double q, double sld_core,double magnetic_sld_core,double eta_core,double radius,
   double sld_solvent,double magnetic_sld_solvent,double eta_solvent,double delta_solvent,
   double fp_n, double sld[],double magnetic_sld[],double eta[], double delta[], double thickness[], double cos_theta, double sin_theta, double alpha, double beta) {
return fqnucsq(q, sld_core, radius, sld_solvent, fp_n, sld, thickness)+fqnuc(q, sld_core, radius, sld_solvent, fp_n, sld, thickness)*fqMz(q, magnetic_sld_core, radius, magnetic_sld_solvent, fp_n, magnetic_sld, thickness,eta_core,eta_solvent,eta)*(1-square_mag_scat(cos_theta, sin_theta,alpha, beta))+fqMzsq(q, magnetic_sld_core, radius, magnetic_sld_solvent, fp_n, magnetic_sld, thickness,eta_core,eta_solvent,eta)*square(1-square_mag_scat(cos_theta, sin_theta,alpha, beta))+fqMtranssq(q, magnetic_sld_core, radius, magnetic_sld_solvent, fp_n, magnetic_sld, thickness,eta_core,eta_solvent,eta, delta_solvent,delta)*(1-square_mag_scat(cos_theta, sin_theta,alpha, beta))*square_mag_scat(cos_theta, sin_theta,alpha, beta);
	}   

//SF Mz^2*(1-rotated_scat_vector[3]^2)*rotated_scat_vector[3]^2 +Mx^2*(1+rotated_scat_vector[3]^4)
static double Idu(double q, double sld_core,double magnetic_sld_core,double eta_core,double radius,
   double sld_solvent,double magnetic_sld_solvent,double eta_solvent,double delta_solvent,
   double fp_n, double sld[],double magnetic_sld[],double eta[], double delta[], double thickness[], double cos_theta, double sin_theta, double alpha, double beta) {
    return fqMtranssq(q, magnetic_sld_core, radius, magnetic_sld_solvent, fp_n, magnetic_sld,thickness,eta_core,eta_solvent,eta, delta_solvent,delta)*(1+square(square_mag_scat(cos_theta, sin_theta,alpha, beta)))+fqMzsq(q, magnetic_sld_core, radius, magnetic_sld_solvent, fp_n,magnetic_sld, thickness,eta_core,eta_solvent,eta)* (1-square_mag_scat(cos_theta, sin_theta,alpha, beta))*square_mag_scat(cos_theta, sin_theta,alpha, beta);
	}   	

static double Iud(double q, double sld_core,double magnetic_sld_core,double eta_core,double radius,
   double sld_solvent,double magnetic_sld_solvent,double eta_solvent,double delta_solvent,
   double fp_n, double sld[],double magnetic_sld[],double eta[], double delta[], double thickness[], double cos_theta, double sin_theta, double alpha, double beta) {
    return Idu(q, sld_core,magnetic_sld_core, eta_core, radius, sld_solvent,magnetic_sld_solvent, eta_solvent,delta_solvent,
   fp_n, sld,magnetic_sld,eta,delta,  thickness, cos_theta, sin_theta, alpha, beta);
	}  



//weighting of spin resolved cross sections to reconstruct partially polarised beam with imperfect optics using up_i/up_f.
static void set_weights(double in_spin, double out_spin, double weight[4]) //from kernel_iq.c
{
  double norm=out_spin;
   
  
  in_spin = clipp(sqrt(square(in_spin)), 0.0, 1.0);//opencl has ambiguities for abs()
  out_spin = clipp(sqrt(square(out_spin)), 0.0, 1.0);

  if (out_spin < 0.5){norm=1-out_spin;}
  


// The norm is needed to make sure that the scattering cross sections are
//correctly weighted, such that the sum of spin-resolved measurements adds up to
// the unpolarised or half-polarised scattering cross section. No intensity weighting
// needed on the incoming polariser side assuming that a user has normalised
// to the incoming flux with polariser in for SANSPOl and unpolarised beam, respectively.


  weight[0] = (1.0-in_spin) * (1.0-out_spin) / norm; // dd
  weight[1] = (1.0-in_spin) * out_spin / norm;       // du
  weight[2] = in_spin * (1.0-out_spin) / norm;       // ud
  weight[3] = in_spin * out_spin / norm;             // uu
}


//calculate 2D from _fq
static double
Iqxy(double qx, double qy, double sld_core,double magnetic_sld_core,double eta_core,double radius,
   double sld_solvent,double magnetic_sld_solvent,double eta_solvent,double delta_solvent,
   double fp_n, double sld[],double magnetic_sld[],double eta[], double delta[], double thickness[], double up_i, double up_f, double alpha, double beta)
{
 const double q = sqrt(qx*qx + qy*qy);
    double cos_theta=qx/q;
    double sin_theta=qy/q;

    double weights[8];  // uu, ud, du, dd, fill,fill, fill, fill (make memory alloc happy)
    set_weights(up_i, up_f, weights);

    
    const double form=weights[0]*Idd(q, sld_core,magnetic_sld_core,eta_core,radius,sld_solvent,magnetic_sld_solvent,eta_solvent,delta_solvent,
   fp_n, sld,magnetic_sld,eta,delta,thickness,cos_theta ,sin_theta, alpha, beta) + weights[1]*Idu(q, sld_core,magnetic_sld_core,eta_core,radius,sld_solvent,magnetic_sld_solvent,eta_solvent,delta_solvent,
   fp_n, sld,magnetic_sld,eta,delta,thickness,cos_theta ,sin_theta, alpha, beta) + weights[2]*Iud(q, sld_core,magnetic_sld_core,eta_core,radius,sld_solvent,magnetic_sld_solvent,eta_solvent,delta_solvent,
   fp_n, sld,magnetic_sld,eta,delta,thickness,cos_theta ,sin_theta, alpha, beta) + weights[3]*Iuu(q, sld_core,magnetic_sld_core,eta_core,radius,sld_solvent,magnetic_sld_solvent,eta_solvent,delta_solvent,
   fp_n, sld,magnetic_sld,eta,delta,thickness,cos_theta ,sin_theta, alpha, beta);
 

   return form;
}



//calculate 1D by averaging over theta
//TODO: choose orientation and sector width for averaging
// 2D to 1D
static double
Iq(double q, double sld_core,double magnetic_sld_core,double eta_core,double radius,
   double sld_solvent,double magnetic_sld_solvent,double eta_solvent,double delta_solvent,
   double fp_n, double sld[],double magnetic_sld[],double eta[], double delta[], double thickness[], double up_i, double up_f, double alpha, double beta)
{
     double weights[8];  // uu, ud, du, dd, fill,fill, fill, fill (make memory alloc happy) 
    set_weights(up_i, up_f, weights);
   double sin_theta, cos_theta; // slots to hold sincos function output of the orientation on the detector plane
  double total_F2 = 0.0;
    for (int i=0; i<GAUSS_N ;i++) {

        const double theta = M_PI * (GAUSS_Z[i] + 1.0); // 0 .. 2 pi
        SINCOS(theta, sin_theta, cos_theta);
        const double form = weights[0]*Idd(q, sld_core,magnetic_sld_core,eta_core,radius,sld_solvent,magnetic_sld_solvent,eta_solvent,delta_solvent,
   fp_n, sld,magnetic_sld,eta,delta,thickness,cos_theta ,sin_theta, alpha, beta) + weights[1]*Idu(q, sld_core,magnetic_sld_core,eta_core,radius,sld_solvent,magnetic_sld_solvent,eta_solvent,delta_solvent,
   fp_n, sld,magnetic_sld,eta,delta,thickness,cos_theta ,sin_theta, alpha, beta) + weights[2]*Iud(q, sld_core,magnetic_sld_core,eta_core,radius,sld_solvent,magnetic_sld_solvent,eta_solvent,delta_solvent,
   fp_n, sld,magnetic_sld,eta,delta,thickness,cos_theta ,sin_theta, alpha, beta) + weights[3]*Iuu(q, sld_core,magnetic_sld_core,eta_core,radius,sld_solvent,magnetic_sld_solvent,eta_solvent,delta_solvent,
   fp_n, sld,magnetic_sld,eta,delta,thickness,cos_theta ,sin_theta, alpha, beta);
       

        total_F2 += GAUSS_W[i] * form ;
    }

    //convert from [1e-12 A-1] to [cm-1]
    return total_F2;
 }






           

File: sasmodels/models/magnetic_langevin_core_shell_3D.py
r"""
Definition
----------
This model describes the SANS of individual (dilute), superparamagnetic particles. The magnetic field is oriented with an inclination of $alpha$ to the neutron beam ans rotated by $beta$.
The model is based on spherical particles with a core multishell structure, see :ref:`core_multi_shell`.
The magnetic scattering length density (SLD) is defined as $\rho_{\mathrm{mag}}=b_H M_S$, 
where $b_H= 2.91*10^{8}A^{-1}m^{-1}$ and $M_S$ is the saturation magnetisation (in $A/m$).

The alignment of the particle moments along the magnetic field is disturbed by thermal fluctuations like Brownian motion of freely rotating particles in a liquid. 
The magnetic SLD $\rho_{\mathrm{mag}}_mag=b_H M_S$ is hence weighted for each magnetisation component with the corresponding average of a Boltzmann orientation distribution 
with $\eta=\mu_0 \mu H/ k_B T$ denotes the Langevin parameter with $\mu_0$ the vacuum permeability and $\mu$ the apparent magnetic moment, and $k_B T$ the thermal energy  [#Wiedenmann2005]_.
The model describes a magnetisation distribution that is equally distributed around the magnetic field direction (no effective prefered easy axis of the ensemble).
In particular, the transversal magnetisation components are randomly oriented at least from one particle to the other.

The resulting magnetisation averages are:

.. math::
    \langle M_{\perp}\rangle=0
    \langle M_{\parallel}\rangle=F_m L(\eta)
    \langle M_{\perp}^2\rangle=F_m^2 L(\eta)/\eta
    \langle M_{\parallel}^2\rangle=F_m^2 (1-2L(\eta)/\eta)

with $F_m$ the magnetic scattering amplitude, the Langevin function $\langle \cos \alpha\rangle= L(\eta)=\coth(\eta)-1/\eta$ and $M_{\perp}$ and $M_\parallel$ the two transversal and the in-field magnetisation components, respectively (for a detailed discussion see [#Muehlbauer2019]_). The transversal magnetisation components of core, shell and solvent are often considered coaligned, corresponding to a coupling parameter $\delta =1$, and 0 if the transversal magnetisation of core and shell are uncorrelated (random orientational relation), i.e. the shell magnetisation component (perpendicular to the field) points in an arbitrary direction with respect to the core spin deflection. The scattering intensity reduces in this case to the sum over the individual squared form factors. The crossed terms $F_{M,i}\cdot F_{M,j}$ disappear as the scattering amplitudes $F_{M,i}$ and $F_{M,j}$ of two different parts can take both signs with equal propability.

The fraction of "upward" neutrons before ('up_frac_i') and after the sample ('up_frac_f') must range between 0 to 1, with 0.5 denoting an unpolarised beam. Note that a fit may result in a negative magnetic SLD, and hence magnetisation, when the polarisation state is inverted, i.e. if you have analysed for a $I_{00}$ state wheras your data are $I_{11}$. The model allows to construct the 4 spin-resolved cross sections (non-spin-flip I_{00}, I_{11} and spin-flip, here I_{01}=I_{10}), half-polarised SANS (SANSpol, incoming polarised beam I_0 and I_1, no analysis after sample 'up_frac_f'$=0.5$), and unpolarised beam ('up_frac_i'$=$'up_frac_f'$=0.5$).
Differences and other combinations between polarised scattering cross section, e.g. to obtain the nuclear-magnetic interference scattering, can be constructed with a custom model (Fitting> Add/Multiply Model) and using approbriate scales.

Every component in the core-multishell-matrix model has a (material dependent) saturation magnetisation and
a Langevin term $\eta$.
The Langevin term (i.e. mean orientation of the magnetic moment to the magnetic field) should be in the simplest case identical for core and shell, and related to the 
average particle moment $\mu=V*M_S$. However, it may vary due to different magnetic anisotropy, spin canting or other disturbances 
acting as a modified effective magnetic field from simple coaligned state.

For dense systems, where a structure factor of the arrangement might be relevant, special care has to be taken to describe 
the local environment around a particle and the structure factor is highly field-dependent due to varying correlations between the particles (see [#Honecker2020]_ for further details). The use of structure model is therefore strongly discouraged. A (micromagnetic) model is often the more approbriate approach.



Validation
----------

The model needs testing and validation.Your feedback is kindly requested.

Demonstrate that the model reproduces known results.

Test correct weighting of spin resolved cross sections to reconstruct unpolarised etc, also with imperfect neutron optics.

Test that 2d gives same result as 1d.



References
----------

.. [#Wiedenmann2005] A. Wiedenmann, *Physica B* 356, 246 (2005)
.. [#Muehlbauer2019] S. Muehlbauer, A. Michels et al., *Rev. Mod. Phys.* 91, 015004 (2019)
.. [#Honecker2020] D. Honecker, L. Fernandez Barguin, and P. Bender, *Phys. Rev. B* 101, 134401 (2020)



Authorship and Verification
----------------------------
* **Author: Dirk Honecker **Date:** September 04, 2020
* **Last Modified by:**
* **Last Reviewed by:**

"""

import numpy as np
from numpy import pi, inf

name = "magnetic_langevin_core_shell_sphere_3D"
title = "Dilute magnetic core-shell particles in a matrix with the magnetisation relaxing with respect to magnetic field"
description = """
    I(q) = A_ij (F_N^2(q)+ C_ij(L(H)) F_N F_M ) +B_ij(L(H)) F_M^2(q) 
            A_ij: weighting function =1 for unpolarised beam and non-neutron-spin-flip scattering, zero for spin-flip scattering
            B_ij(L(H)): weighting function for purely magnetic scattering, different for the various possible spin-resolved scattering cross sections
            C_ij(L(H)): weighting function for nuclear-magnetic interference scattering
            L(H): Langevin term describing the orientational distribution of magnetic moments with respect to magnetic field
            F_N: nuclear form factor
            F_M: magnetic form factor
The core and each shell can have a unique thickness, Langevin parameter and magnetic and nuclear sld.
"""
category = "shape:sphere"

# pylint: disable=bad-whitespace, line-too-long
#             ["name", "units", default, [lower, upper], "type","description"],
parameters = [["nuc_sld_core", "1e-6/Ang^2", 1.0, [-inf, inf], "", "Core scattering length density"],
              ["magnetic_sld_core",    "1e-6/Ang^2", 1.0,  [-inf, inf], "",    "Magnetic core scattering length density"],
              ["eta_core", "None", 3.0,  [0, inf], "",    "Langevin parameter of core"],
              ["radius", "Ang", 50., [0, inf], "volume", "Radius of the core"],
              ["nuc_sld_solvent", "1e-6/Ang^2", 6.4, [-inf, inf], "", "Solvent scattering length density"],
              ["magnetic_sld_solvent", "1e-6/Ang^2", 3.0,  [-inf, inf], "",    "Magnetic Solvent scattering length density"],
              ["eta_solvent", "None ", 3.0,  [0, inf], "",    "Langevin parameter of solvent of solvent"],
              ["delta_solvent", "None", 1.0,  [0, 1], "",    "Disorder coupling parameter of matrix to core"],
              ["n", "", 1, [0, 10], "volume", "Number of shells"],
              ["nuc_sld_shell[n]", "1e-6/Ang^2", 1.7, [-inf, inf], "", "Scattering length density of shell k"],
              ["magnetic_sld_shell[n]", "1e-6/Ang^2", 1.7, [-inf, inf], "", "Magnetic scattering length density of shell k"],
              ["eta[n]", "None", 3.0,  [0, inf], "",    "Langevin parameter of shell k"],
              ["delta[n]", "None", 1.0,  [0, 1], "",    "Disorder coupling parameter of shell to core"],
              ["thickness[n]", "Ang", 40., [0, inf], "volume", "Thickness of shell k"],
              ["up_i",          "None",  0, [0, 1], "", "Polarisation incoming beam"],
              ["up_f",          "None",  0, [0, 1], "", "Polarisation outgoing beam"],
              ["alpha",          "None",  0, [0, 180], "", "inclination of field to neutron beam"],
              ["beta",          "None",  0, [0, 360], "", "rotation of field around neutron beam"],             
             ]
# pylint: enable=bad-whitespace, line-too-long




source = ["lib/sas_3j1x_x.c","lib/gauss76.c", "magnetic_langevin_core_shell_3D.c"]
structure_factor = False
have_Fq = False
single=False


effective_radius_type = ["outer radius", "core radius"]

def random():
    """Return a random parameter set for the model."""
    num_shells = np.minimum(np.random.poisson(3)+1, 10)
    total_radius = 10**np.random.uniform(1.7, 4)
    thickness = np.random.exponential(size=num_shells+1)
    thickness *= total_radius/np.sum(thickness)
    pars = dict(
        #background=0,
        n=num_shells,
        radius=thickness[0],
    )
    for k, v in enumerate(thickness[1:]):
        pars['thickness%d'%(k+1)] = v
    return pars

def profile(nuc_sld_core, magnetic_sld_core,eta_core,radius, nuc_sld_solvent,magnetic_sld_solvent,eta_solvent,delta_solvent, n, nuc_sld_shell,magnetic_sld_shell,eta,delta, thickness,up_i,up_f, alpha,beta):
    """
    Returns the SLD profile *r* (Ang), and *rho* (1e-6/Ang^2).
    """
    n = int(n+0.5)
    z = []
    rho = []
    rho_mag = []

    # add in the core
    z.append(0)
    rho.append(nuc_sld_core)
    rho_mag.append(magnetic_sld_core)
    z.append(radius)
    rho.append(nuc_sld_core)
    rho_mag.append(magnetic_sld_core)

    # add in the shells
    for k in range(int(n)):
        # Left side of each shells
        z.append(z[-1])
        rho.append(nuc_sld_shell[k])
        rho_mag.append(magnetic_sld_shell[k])
        z.append(z[-1] + thickness[k])
        rho.append(nuc_sld_shell[k])
    # add in the solvent
    z.append(z[-1])
    rho.append(nuc_sld_solvent)
    rho_mag.append(magnetic_sld_solvent)
    z.append(z[-1]*1.25)
    rho.append(nuc_sld_solvent)
    rho_mag.append(magnetic_sld_solvent)

    return np.asarray(z), np.asarray(rho)
    







# tests = [
    # [{'radius': 20.0, 'thickness': 10.0}, 0.1, None, None, 30.0, 4.*pi/3*30**3, 1.0],

    # # The SasView test result was 0.00169, with a background of 0.001
    # [{'radius': 60.0, 'thickness': 10.0, 'sld_core': 1.0, 'sld_shell': 2.0,
      # 'sld_solvent': 3.0, 'background': 0.0}, 0.4, 0.000698838],
# ]
